File: java/src/main/java/com/learning/bigoanalysis/ex6OptimizeCode.java
package com.learning.bigoanalysis;

import java.util.*;

/**
 * Exercise 6: Optimize Code Complexity (Hard)
 *
 * Each method below has an inefficient implementation. Your task is to:
 * 1. Analyze the current time complexity
 * 2. Implement an optimized version with better complexity
 * 3. Return the complexity of your optimized solution
 *
 * The original inefficient implementations are provided for reference.
 *
 * Difficulty: Hard
 * Topics: Optimization, Algorithm Design, Trade-offs
 */
public class ex6OptimizeCode {

    /**
     * Problem 1: Check if array contains duplicates
     * Current: O(n²) with nested loops
     * Goal: Optimize to O(n)
     */
    public static boolean hasDuplicatesOptimized(int[] arr) {
        // TODO: Implement an O(n) solution
        // Hint: Use a HashSet to track seen elements
        throw new UnsupportedOperationException("Not implemented yet");
    }

    public static String getComplexityHasDuplicates() {
        // TODO: Return the complexity of your optimized solution
        throw new UnsupportedOperationException("Not implemented yet");
    }

    /**
     * Problem 2: Find two numbers that sum to target
     * Current: O(n²) with nested loops
     * Goal: Optimize to O(n)
     */
    public static int[] twoSumOptimized(int[] arr, int target) {
        // TODO: Implement an O(n) solution
        // Hint: Use a HashMap to store complements
        // Return indices of the two numbers, or {-1, -1} if not found
        throw new UnsupportedOperationException("Not implemented yet");
    }

    public static String getComplexityTwoSum() {
        // TODO: Return the complexity of your optimized solution
        throw new UnsupportedOperationException("Not implemented yet");
    }

    /**
     * Problem 3: Find first non-repeating character
     * Current: O(n²) checking count for each character
     * Goal: Optimize to O(n)
     */
    public static char firstNonRepeatingOptimized(String s) {
        // TODO: Implement an O(n) solution
        // Hint: Use a HashMap to count frequencies in one pass,
        // then find the first character with count 1 in a second pass
        // Return '\0' if no non-repeating character exists
        throw new UnsupportedOperationException("Not implemented yet");
    }

    public static String getComplexityFirstNonRepeating() {
        // TODO: Return the complexity of your optimized solution
        throw new UnsupportedOperationException("Not implemented yet");
    }

    /**
     * Problem 4: Find common elements in two sorted arrays
     * Current: O(n*m) with nested loops
     * Goal: Optimize to O(n + m)
     */
    public static List<Integer> findCommonOptimized(int[] arr1, int[] arr2) {
        // TODO: Implement an O(n + m) solution
        // Hint: Use two pointers since arrays are sorted
        throw new UnsupportedOperationException("Not implemented yet");
    }

    public static String getComplexityFindCommon() {
        // TODO: Return the complexity of your optimized solution
        // Express as O(n + m) where n and m are the array sizes
        throw new UnsupportedOperationException("Not implemented yet");
    }

    /**
     * Problem 5: Compute nth Fibonacci number
     * Current: O(2^n) with naive recursion
     * Goal: Optimize to O(n)
     */
    public static long fibonacciOptimized(int n) {
        // TODO: Implement an O(n) solution
        // Hint: Use dynamic programming (iterative or memoization)
        throw new UnsupportedOperationException("Not implemented yet");
    }

    public static String getComplexityFibonacci() {
        // TODO: Return the complexity of your optimized solution
        throw new UnsupportedOperationException("Not implemented yet");
    }

    /**
     * Problem 6: Find subarray with maximum sum
     * Current: O(n²) or O(n³) with brute force
     * Goal: Optimize to O(n) using Kadane's algorithm
     */
    public static int maxSubarraySumOptimized(int[] arr) {
        // TODO: Implement Kadane's algorithm - O(n)
        // Hint: Track current sum and maximum sum seen so far
        throw new UnsupportedOperationException("Not implemented yet");
    }

    public static String getComplexityMaxSubarraySum() {
        // TODO: Return the complexity of your optimized solution
        throw new UnsupportedOperationException("Not implemented yet");
    }

    // Reference inefficient implementations (for comparison)
    @SuppressWarnings("unused")
    private static class InefficientVersions {
        // O(n²) - has duplicates
        static boolean hasDuplicatesInefficient(int[] arr) {
            for (int i = 0; i < arr.length; i++) {
                for (int j = i + 1; j < arr.length; j++) {
                    if (arr[i] == arr[j]) return true;
                }
            }
            return false;
        }

        // O(n²) - two sum
        static int[] twoSumInefficient(int[] arr, int target) {
            for (int i = 0; i < arr.length; i++) {
                for (int j = i + 1; j < arr.length; j++) {
                    if (arr[i] + arr[j] == target) {
                        return new int[]{i, j};
                    }
                }
            }
            return new int[]{-1, -1};
        }

        // O(n²) - first non-repeating
        static char firstNonRepeatingInefficient(String s) {
            for (int i = 0; i < s.length(); i++) {
                boolean isUnique = true;
                for (int j = 0; j < s.length(); j++) {
                    if (i != j && s.charAt(i) == s.charAt(j)) {
                        isUnique = false;
                        break;
                    }
                }
                if (isUnique) return s.charAt(i);
            }
            return '\0';
        }

        // O(2^n) - fibonacci
        static long fibonacciInefficient(int n) {
            if (n <= 1) return n;
            return fibonacciInefficient(n - 1) + fibonacciInefficient(n - 2);
        }

        // O(n²) - max subarray sum
        static int maxSubarraySumInefficient(int[] arr) {
            int maxSum = Integer.MIN_VALUE;
            for (int i = 0; i < arr.length; i++) {
                int currentSum = 0;
                for (int j = i; j < arr.length; j++) {
                    currentSum += arr[j];
                    maxSum = Math.max(maxSum, currentSum);
                }
            }
            return maxSum;
        }
    }

    public static void main(String[] args) {
        System.out.println("Big O Analysis - Exercise 6: Optimize Code");
        System.out.println("Implement optimized versions of inefficient algorithms\n");

        try {
            System.out.println("Has Duplicates: " + getComplexityHasDuplicates());
            System.out.println("Two Sum: " + getComplexityTwoSum());
            System.out.println("First Non-Repeating: " + getComplexityFirstNonRepeating());
            System.out.println("Find Common: " + getComplexityFindCommon());
            System.out.println("Fibonacci: " + getComplexityFibonacci());
            System.out.println("Max Subarray Sum: " + getComplexityMaxSubarraySum());
        } catch (UnsupportedOperationException e) {
            System.out.println("Not yet implemented!");
        }
    }
}

File: python/big_o_analysis/ex6_optimize_code.py
"""
Exercise 6: Optimize Code Complexity (Hard)

Each function below has an inefficient implementation. Your task is to:
1. Analyze the current time complexity
2. Implement an optimized version with better complexity
3. Return the complexity of your optimized solution

The original inefficient implementations are provided for reference.

Difficulty: Hard
Topics: Optimization, Algorithm Design, Trade-offs
"""

from typing import List, Tuple


def has_duplicates_optimized(arr: List[int]) -> bool:
    """
    Problem 1: Check if array contains duplicates
    Current: O(n²) with nested loops
    Goal: Optimize to O(n)

    Hint: Use a set to track seen elements
    """
    # TODO: Implement an O(n) solution
    raise NotImplementedError("Not implemented yet")


def get_complexity_has_duplicates() -> str:
    """Return the complexity of your optimized solution"""
    # TODO: Implement this
    raise NotImplementedError("Not implemented yet")


def two_sum_optimized(arr: List[int], target: int) -> Tuple[int, int]:
    """
    Problem 2: Find two numbers that sum to target
    Current: O(n²) with nested loops
    Goal: Optimize to O(n)

    Hint: Use a dictionary to store complements
    Return indices of the two numbers, or (-1, -1) if not found
    """
    # TODO: Implement an O(n) solution
    raise NotImplementedError("Not implemented yet")


def get_complexity_two_sum() -> str:
    """Return the complexity of your optimized solution"""
    # TODO: Implement this
    raise NotImplementedError("Not implemented yet")


def first_non_repeating_optimized(s: str) -> str:
    """
    Problem 3: Find first non-repeating character
    Current: O(n²) checking count for each character
    Goal: Optimize to O(n)

    Hint: Use a dictionary to count frequencies in one pass,
    then find the first character with count 1 in a second pass
    Return empty string if no non-repeating character exists
    """
    # TODO: Implement an O(n) solution
    raise NotImplementedError("Not implemented yet")


def get_complexity_first_non_repeating() -> str:
    """Return the complexity of your optimized solution"""
    # TODO: Implement this
    raise NotImplementedError("Not implemented yet")


def find_common_optimized(arr1: List[int], arr2: List[int]) -> List[int]:
    """
    Problem 4: Find common elements in two sorted arrays
    Current: O(n*m) with nested loops
    Goal: Optimize to O(n + m)

    Hint: Use two pointers since arrays are sorted
    """
    # TODO: Implement an O(n + m) solution
    raise NotImplementedError("Not implemented yet")


def get_complexity_find_common() -> str:
    """
    Return the complexity of your optimized solution.
    Express as O(n + m) where n and m are the array sizes
    """
    # TODO: Implement this
    raise NotImplementedError("Not implemented yet")


def fibonacci_optimized(n: int) -> int:
    """
    Problem 5: Compute nth Fibonacci number
    Current: O(2^n) with naive recursion
    Goal: Optimize to O(n)

    Hint: Use dynamic programming (iterative or memoization)
    """
    # TODO: Implement an O(n) solution
    raise NotImplementedError("Not implemented yet")


def get_complexity_fibonacci() -> str:
    """Return the complexity of your optimized solution"""
    # TODO: Implement this
    raise NotImplementedError("Not implemented yet")


def max_subarray_sum_optimized(arr: List[int]) -> int:
    """
    Problem 6: Find subarray with maximum sum
    Current: O(n²) or O(n³) with brute force
    Goal: Optimize to O(n) using Kadane's algorithm

    Hint: Track current sum and maximum sum seen so far
    """
    # TODO: Implement Kadane's algorithm - O(n)
    raise NotImplementedError("Not implemented yet")


def get_complexity_max_subarray_sum() -> str:
    """Return the complexity of your optimized solution"""
    # TODO: Implement this
    raise NotImplementedError("Not implemented yet")


# Reference inefficient implementations (for comparison)
class InefficientVersions:
    """Reference inefficient implementations for comparison"""

    @staticmethod
    def has_duplicates_inefficient(arr: List[int]) -> bool:
        """O(n²) - has duplicates"""
        for i in range(len(arr)):
            for j in range(i + 1, len(arr)):
                if arr[i] == arr[j]:
                    return True
        return False

    @staticmethod
    def two_sum_inefficient(arr: List[int], target: int) -> Tuple[int, int]:
        """O(n²) - two sum"""
        for i in range(len(arr)):
            for j in range(i + 1, len(arr)):
                if arr[i] + arr[j] == target:
                    return (i, j)
        return (-1, -1)

    @staticmethod
    def first_non_repeating_inefficient(s: str) -> str:
        """O(n²) - first non-repeating"""
        for i in range(len(s)):
            is_unique = True
            for j in range(len(s)):
                if i != j and s[i] == s[j]:
                    is_unique = False
                    break
            if is_unique:
                return s[i]
        return ""

    @staticmethod
    def fibonacci_inefficient(n: int) -> int:
        """O(2^n) - fibonacci"""
        if n <= 1:
            return n
        return InefficientVersions.fibonacci_inefficient(n - 1) + \
               InefficientVersions.fibonacci_inefficient(n - 2)

    @staticmethod
    def max_subarray_sum_inefficient(arr: List[int]) -> int:
        """O(n²) - max subarray sum"""
        max_sum = float('-inf')
        for i in range(len(arr)):
            current_sum = 0
            for j in range(i, len(arr)):
                current_sum += arr[j]
                max_sum = max(max_sum, current_sum)
        return int(max_sum)


if __name__ == "__main__":
    print("Big O Analysis - Exercise 6: Optimize Code")
    print("Implement optimized versions of inefficient algorithms\n")

    try:
        print(f"Has Duplicates: {get_complexity_has_duplicates()}")
        print(f"Two Sum: {get_complexity_two_sum()}")
        print(f"First Non-Repeating: {get_complexity_first_non_repeating()}")
        print(f"Find Common: {get_complexity_find_common()}")
        print(f"Fibonacci: {get_complexity_fibonacci()}")
        print(f"Max Subarray Sum: {get_complexity_max_subarray_sum()}")
    except NotImplementedError:
        print("Not yet implemented!")
